plugins/manager/manager.go

```diff
@@ -21,121 +21,76 @@ import (
+	sink          metrics.DataSink
+	pushInterval  time.Duration
+	ticker        *time.Ticker
+	stopChan      chan struct{}
 }
 
-func NewManager(source metrics.MetricsSource, processors []metrics.DataProcessor, sink metrics.DataSink, resolution time.Duration,
-	scrapeOffset time.Duration, maxParallelism int) (Manager, error) {
-	manager := realManager{
-		source:                 source,
-		processors:             processors,
-		sink:                   sink,
-		resolution:             resolution,
-		scrapeOffset:           scrapeOffset,
-		stopChan:               make(chan struct{}),
-		housekeepSemaphoreChan: make(chan struct{}, maxParallelism),
-		housekeepTimeout:       resolution / 2,
-	}
-
-	for i := 0; i < maxParallelism; i++ {
-		manager.housekeepSemaphoreChan <- struct{}{}
+// NewPushManager crates a new PushManager
+func NewPushManager(source sources.SourceManager, processors []metrics.DataProcessor,
+	sink metrics.DataSink, pushInterval time.Duration) (PushManager, error) {
+	manager := pushManagerImpl{
+		sourceManager: source,
+		processors:    processors,
+		sink:          sink,
+		pushInterval:  pushInterval,
+		stopChan:      make(chan struct{}),
 	}
 
 	return &manager, nil
 }
 
-func (rm *realManager) Start() {
-	go rm.Housekeep()
-}
-
-func (rm *realManager) Stop() {
-	rm.stopChan <- struct{}{}
-}
-
-func (rm *realManager) Handle(cfg interface{}) {
-	// no-op
+func (rm *pushManagerImpl) Start() {
+	rm.ticker = time.NewTicker(rm.pushInterval)
+	go rm.run()
 }
 
-func (rm *realManager) Housekeep() {
+func (rm *pushManagerImpl) run() {
 	for {
-		// Always try to get the newest metrics
-		now := time.Now()
-		start := now.Truncate(rm.resolution)
-		end := start.Add(rm.resolution)
-		timeToNextSync := end.Add(rm.scrapeOffset).Sub(now)
-
 		select {
-		case <-time.After(timeToNextSync):
-			rm.housekeep(start, end)
+		case <-rm.ticker.C:
+			go rm.push()
 		case <-rm.stopChan:
+			rm.ticker.Stop()
+			rm.sourceManager.Stop()
 			rm.sink.Stop()
 			return
 		}
 	}
 }
 
-func (rm *realManager) housekeep(start, end time.Time) {
-	if !start.Before(end) {
-		glog.Warningf("Wrong time provided to housekeep start:%s end: %s", start, end)
-		return
-	}
-
-	select {
-	case <-rm.housekeepSemaphoreChan:
-		// ok, good to go
-
-	case <-time.After(rm.housekeepTimeout):
-		glog.Warningf("Spent too long waiting for housekeeping to start")
-		return
-	}
-
-	go func(rm *realManager) {
-		// should always give back the semaphore
-		defer func() { rm.housekeepSemaphoreChan <- struct{}{} }()
-		data, err := rm.source.ScrapeMetrics(start, end)
-
-		if err != nil {
-			glog.Errorf("Error in scraping metrics for %s: %v", rm.source.Name(), err)
-			return
-		}
+func (rm *pushManagerImpl) Stop() {
+	rm.stopChan <- struct{}{}
+}
 
+func (rm *pushManagerImpl) push() {
+	dataList := rm.sourceManager.GetPendingMetrics()
+	for _, data := range dataList {
 		for _, p := range rm.processors {
-			newData, err := process(p, data)
+			newData, err := p.Process(data)
```

Comment:
All the processors we have today work on `batch.MetricSets`. Let's just check if that is not nil and only then invoke `Process()`.